reflection/serverreflection.go

```diff
@@ -65,291 +60,165 @@ type GRPCServer interface {
 	rpb.UnimplementedServerReflectionServer
-	s GRPCServer
+	s            GRPCServer
+	descResolver protodesc.Resolver
+	extResolver  ExtensionResolver
 
-	initSymbols  sync.Once
-	serviceNames []string
-	symbols      map[string]*dpb.FileDescriptorProto // map of fully-qualified names to files
+	initServiceNames sync.Once
+	serviceNames     []string
 }
 
-// Register registers the server reflection service on the given gRPC server.
-func Register(s GRPCServer) {
-	rpb.RegisterServerReflectionServer(s, &serverReflectionServer{
-		s: s,
-	})
+// ServerOptions represents the options used to construct a reflection server.
+//
+// Either Server or ServiceNames must be populated, but not both. These control
+// what services are advertised by the server in the ListServices capability of
+// the reflection service. If neither is provided, the returned server can still
+// server descriptors, but it will advertise no service names.
+//
+// The given DescriptorResolver will be used to resolve symbols and files by
+// name. If not present, protoregistry.GlobalFiles will be used. The given
+// ExtensionResolver will be used to resolve extensions. If not present,
+// protoregistry.GlobalTypes will be used.
+type ServerOptions struct {
+	// An RPC server, whose exposed services are made available via service
+	// reflection.
+	Server GRPCServer
+	// The list of service names. This should only be populated if Server is
+	// nil.
+	ServiceNames []string
+	// Optional resolver used to load descriptors.
+	DescriptorResolver protodesc.Resolver
+	// Optional resolver used to query for known extensions.
+	ExtensionResolver ExtensionResolver
 }
 
-func (s *serverReflectionServer) getSymbols() (svcNames []string, symbolIndex map[string]*dpb.FileDescriptorProto) {
-	s.initSymbols.Do(func() {
-		serviceInfo := s.s.GetServiceInfo()
-
-		s.symbols = map[string]*dpb.FileDescriptorProto{}
-		s.serviceNames = make([]string, 0, len(serviceInfo))
-		processed := map[string]struct{}{}
-		for svc, info := range serviceInfo {
-			s.serviceNames = append(s.serviceNames, svc)
-			fdenc, ok := parseMetadata(info.Metadata)
-			if !ok {
-				continue
-			}
-			fd, err := decodeFileDesc(fdenc)
-			if err != nil {
-				continue
-			}
-			s.processFile(fd, processed)
-		}
-		sort.Strings(s.serviceNames)
-	})
-
-	return s.serviceNames, s.symbols
-}
-
-func (s *serverReflectionServer) processFile(fd *dpb.FileDescriptorProto, processed map[string]struct{}) {
-	filename := fd.GetName()
-	if _, ok := processed[filename]; ok {
-		return
-	}
-	processed[filename] = struct{}{}
-
-	prefix := fd.GetPackage()
-
-	for _, msg := range fd.MessageType {
-		s.processMessage(fd, prefix, msg)
-	}
-	for _, en := range fd.EnumType {
-		s.processEnum(fd, prefix, en)
+// NewServer returns a reflection server implementation using the given options.
+// It returns an error if the given options are invalid.
+func NewServer(opts ServerOptions) (rpb.ServerReflectionServer, error) {
```

Comment:
I'm curious what your thoughts on this style of API, vs. `Register` where the server object is never exposed.

One advantage here is that it allows the caller to wrap/decorate the service. (Other formulation requires caller to use an interceptor to do so.)

Also, the options include an optional `GRPCServer`. But we'd _have_ to have a server to do the actual `Register...` call. So it feels like a confusing API to have a required server argument and then also an optional field in the options struct.

Reply:
Or, we define an interface with `func GetServiceInfo()` (and you won't need the other field `ServiceNames []string`).

Did we talk about this interface option before? I don't remember if we got any conclusion.

Reply:
There is already an interface named `GRPCServer`. The list of `ServiceNames` is useful if the caller wants more control over exactly _which_ services are advertised by the reflection service.

Reply:
Oh, right. I forgot we already made the `GRPCServer` interface.

What I meant is a smaller interface, with just `GetServiceInfo() map[string]grpc.ServiceInfo`.
You can still pass it the grpc server (that's the default behavior), or the dynamic factory.
Or if you want a static `[]string`, you can pass it an implementation that returns the list (and you won't need the other field).

Reply:
Okay, I think I see what you mean.

Would you object to me adding a utility function, like `ServiceInfoFromNames`, that takes a slice of names (`[]string`) and returns this new interface, that provides the service info?

Reply:
Done. (I went ahead and added the utility function I asked about. Let me know what you think.)